service/src/LevelZero.cpp

```diff
@@ -509,48 +543,95 @@ namespace geopm
         return {result_active, result_timestamp};
     }
 
-    uint64_t LevelZeroImp::energy_timestamp(unsigned int l0_device_idx) const
+    uint64_t LevelZeroImp::energy_timestamp(int geopm_domain,
+                                            unsigned int l0_device_idx,
+                                            int l0_domain,
+                                            int l0_domain_idx) const
     {
-        return m_devices.at(l0_device_idx).cached_energy_timestamp;
+        uint64_t timestamp = 0;
+        //PACKAGE
+        if(geopm_domain == GEOPM_DOMAIN_GPU) {
+            timestamp = m_devices.at(l0_device_idx).cached_energy_timestamp;
+        }
+        //TILE
+        else if(geopm_domain == GEOPM_DOMAIN_GPU_CHIP) {
+            timestamp = m_devices.at(l0_device_idx).subdevice.cached_energy_timestamp.at(l0_domain_idx);
+        }
+        return timestamp;
     }
 
-    uint64_t LevelZeroImp::energy(unsigned int l0_device_idx) const
+    uint64_t LevelZeroImp::energy(int geopm_domain, unsigned int l0_device_idx,
+                                  int l0_domain, int l0_domain_idx) const
     {
-        return energy_pair(l0_device_idx).first;
+        return energy_pair(geopm_domain, l0_device_idx, l0_domain_idx).first;
     }
 
-    std::pair<uint64_t,uint64_t> LevelZeroImp::energy_pair(unsigned int l0_device_idx) const
+    std::pair<uint64_t,uint64_t> LevelZeroImp::energy_pair(int geopm_domain,
+                                                           unsigned int l0_device_idx,
+                                                           int l0_domain_idx) const
     {
         ze_result_t ze_result;
```

Comment:
I'm wondering now if we should just initialize all structs with {} that are ever pass-by-reference in a new PR...